### PR TITLE
Adding Southside encounters

### PR DESCRIPTION
**Pull request by ThomasLipp**
Adding Southside encounters, and also fixing some references to stack instead of event_stack.

### CODE REVIEW

File: eldritch/test_encounters.py [hunk 1]
```diff
@@ -520,6 +624,343 @@ def testSociety4Fail(self):
...
+    self.assertEqual(len(self.state.common), 1)
+
+  def testHouse6Fail(self):
+    self.state.event_stack.append(encounters.House6(self.char))
+    with mock.patch.object(events.random, "randint", new=mock.MagicMock(return_value=2)):
+      self.resolve_until_done()
+    self.assertEqual(self.char.stamina, 3)
+    self.assertEqual(self.char.sanity, 3)
+
+  def testHouse6PassSanity(self):
```
Comment (fingoldfish) [inline eldritch/test_encounters.py:741]: See comments above - this should be a loss.

File: eldritch/test_encounters.py [hunk 2]
```diff
@@ -520,6 +624,343 @@ def testSociety4Fail(self):
...
+    choice = self.resolve_to_choice(MultipleChoice)
+    choice.resolve(self.state,"Yes")
+    choice2 = self.resolve_to_choice(MultipleChoice)
+    choice2.resolve(self.state, 3)
+    self.resolve_until_done()
+    self.assertEqual(self.char.dollars, 2)
+    self.assertEqual(self.char.stamina, 2)
+    self.assertEqual(self.char.sanity, 4)
+
+  def testHouse7Yes4(self):
```
Comment (fingoldfish) [inline eldritch/test_encounters.py:833]: You don't actually have to test each choice here - the unit tests for SplitGain should be enough.

File: eldritch/test_encounters.py [hunk 3]
```diff
@@ -520,6 +624,343 @@ def testSociety4Fail(self):
...
+    self.resolve_until_done()
+    self.assertEqual(self.char.bless_curse, 1)
+
+  def testChurch2Curse(self):
+    self.char.bless_curse = -1
+    self.state.event_stack.append(encounters.Church2(self.char))
+    self.resolve_until_done()
+    self.assertEqual(self.char.bless_curse, 0)
+
+  def testChurch3money(self):
```
Comment (fingoldfish) [inline eldritch/test_encounters.py:877]: Please capitalize the M in Money

File: eldritch/test_encounters.py [hunk 4]
```diff
@@ -520,6 +624,343 @@ def testSociety4Fail(self):
...
+  def testChurch6CluesNo(self):
+    self.state.event_stack.append(encounters.Church6(self.char))
+    self.char.clues = 2
+    self.assertEqual(self.char.clues, 2)
+    choice = self.resolve_to_choice(MultipleChoice)
+    choice.resolve(self.state, "No")
+    self.resolve_until_done()
+    self.assertEqual(self.char.clues, 2)
+
+  def testChurch6CluesNo(self):
```
Comment (fingoldfish) [inline eldritch/test_encounters.py:940]: You seem to have accidentally named this the same as the previous test.

**Review by fingoldfish — CHANGES_REQUESTED** [review]

File: eldritch/encounters.py [hunk 5]
```diff
@@ -252,11 +252,134 @@ def Graveyard6(char):
...
+  move_dream = events.ForceMovement(char, "Dreamlands1")
+  # TODO: don't hard-code.
+  enc = events.GateEncounter(char, "Dreamlands", {"blue", "green", "red", "yellow"})
+  ret = events.ForceMovement(char, "Society")
+  dreamlands =  events.Sequence([move_dream, enc, ret], char)
+  spell = events.Draw(char, "spells", 1)
+  pay = events.Loss(char, {"dollars": 3})
+  searchstacks = events.PassFail(char, luck, spell, dreamlands)
+  library = events.Sequence([pay, searchstacks], char)
+  choice = events.BinaryChoice(char,"Pay $3 to access the private library", "Yes", "No", library, move)
```
Comment (fingoldfish) [inline eldritch/encounters.py:273]: BinaryChoice can now accept a prerequisite so you no longer need to use PassFail below. This changes the unit tests slightly in that you must now use resolve_to_choice(). See Diner3 for an example.

File: eldritch/encounters.py [hunk 6]
```diff
@@ -252,11 +252,134 @@ def Graveyard6(char):
...
   skill = events.Sequence([events.Draw(char, "skills", 1), events.Delayed(char)], char)
   cond = events.Conditional(char, check, "successes", {0: events.Nothing(), 2: skill})
   return events.Sequence([check, cond], char)
+def Society5(char):
+  return events.Loss(char, {"sanity": 1})
+def Society6(char):
+  # TODO: spend gate trophy
+  prereq = events.AttributePrerequisite(char, "dollars", 1, "at least") # TODO: actually a gate trophy
+  # TODO: prerquisite on ally
+  ally = events.Draw(char, "allies", "Old Professor")
```
Comment (fingoldfish) [inline eldritch/encounters.py:293]: Can now be replaced with GainAllyOrReward.

File: eldritch/encounters.py [hunk 7]
```diff
@@ -252,11 +252,134 @@ def Graveyard6(char):
...
+  stamina = events.Gain(char, {"stamina": 1})
+  sanity = events.Gain(char, {"sanity": 1})
+  choice = events.BinaryChoice(char, "Gain 1 stamina or sanity?", "Stamina", "Sanity", stamina, sanity)
+  return events.PassFail(char, luck, choice, events.Nothing())
+def House7(char):
+  prereq = events.AttributePrerequisite(char, "dollars", 3, "at least")
+  pay = events.Loss(char, {"dollars": 3})
+  gain = events.SplitGain(char, "sanity", "stamina", 4)
+  stay = events.Sequence([pay, gain], char)
+  choice = events.BinaryChoice(char, "Spend $3 to spend the night?", "Yes", "No", stay, events.Nothing())
```
Comment (fingoldfish) [inline eldritch/encounters.py:343]: Same thing - can be converted to a BinaryChoice with a prerequisite.

File: eldritch/encounters.py [hunk 8]
```diff
@@ -252,11 +252,134 @@ def Graveyard6(char):
...
+  clue = events.Loss(char, {"clues": 1})
+  roll = events.Nothing() #TODO: roll for a success
+  doom = events.Nothing() #TODO: remove a doom token
+  chance = events.Sequence([clue, roll, doom], char)
+  choice = events.BinaryChoice(char, "Spend a clue token for a chance to remove a Doom Token?", 
+                               "Yes",
+                               "No",
+                               chance,
+                               events.Nothing())
+  return events.PassFail(char, prereq, choice, events.Nothing())
```
Comment (fingoldfish) [inline eldritch/encounters.py:376]: Also can change this to adding a prerequisite to the BinaryChoice.

File: eldritch/encounters.py [hunk 9]
```diff
@@ -252,11 +252,134 @@ def Graveyard6(char):
...
+  converse = events.PassFail(char, will, item, stay)
+  return events.BinaryChoice(char, "Converse with travelling Salesman?", "Yes", "No", converse, events.Nothing())
+def House5(char):
+  return events.Draw(char, "common", 1)
+def House6(char):
+  luck = events.Check(char, "luck", -1)
+  stamina = events.Gain(char, {"stamina": 1})
+  sanity = events.Gain(char, {"sanity": 1})
+  choice = events.BinaryChoice(char, "Gain 1 stamina or sanity?", "Stamina", "Sanity", stamina, sanity)
+  return events.PassFail(char, luck, choice, events.Nothing())
```
[the remark this answers is not in the document]
Comment (ThomasLipp) [reply eldritch/encounters.py:337]: wishful thinking?

File: eldritch/encounters.py [hunk 10]
```diff
@@ -252,11 +252,134 @@ def Graveyard6(char):
...
+  cond =  events.Conditional(char, check, "successes", {0: loseSanity, 1: move, 2: gainSanity})
+  return events.Sequence([check, cond], char)
+def Church6(char):
+  #TODO: remove a counter from the doomtrack
+  prereq = events.AttributePrerequisite(char, "clues", 1, "at least")
+  clue = events.Loss(char, {"clues": 1})
+  roll = events.Nothing() #TODO: roll for a success
+  doom = events.Nothing() #TODO: remove a doom token
+  chance = events.Sequence([clue, roll, doom], char)
+  choice = events.BinaryChoice(char, "Spend a clue token for a chance to remove a Doom Token?",
```
[the remark this answers is not in the document]
Comment (ThomasLipp) [reply eldritch/encounters.py:371]: That will take some adjustment, definitely not the preferred way at work, but will try to do that here

File: eldritch/encounters.py [hunk 11]
```diff
@@ -252,11 +252,134 @@ def Graveyard6(char):
...
+  items = events.Nothing() # TODO: Lose half your items
+  return events.BinaryChoice(char, "Donate half or your money or half of your items.","Money", "Items", money, items)
+def Church4(char):
+  holywater = events.DrawSpecific(char, "unique", "Holy Water")
+  return events.BinaryChoice(char, "Search for Holy Water?", "Yes", "No", holywater, events.Nothing())
+def Church5(char):
+  check = events.Check(char, "luck", 0)
+  loseSanity = events.Loss(char, {"sanity": 3})
+  move = events.ForceMovement(char, "Southside")
+  gainSanity = events.Gain(char, {"sanity": char.max_sanity})
```
[the remark this answers is not in the document]
Comment (ThomasLipp) [reply eldritch/encounters.py:361]: I may have missed the style guide.  Would you prefer GainSanity or gain_sanity? (probably not going to change it, but good to know for the future)

File: eldritch/encounters.py [hunk 12]
```diff
@@ -252,11 +252,134 @@ def Graveyard6(char):
...
+def Church3(char):
+  money = events.Loss(char, {"dollars": (char.dollars + 1)//2})
+  items = events.Nothing() # TODO: Lose half your items
+  return events.BinaryChoice(char, "Donate half or your money or half of your items.","Money", "Items", money, items)
+def Church4(char):
+  holywater = events.DrawSpecific(char, "unique", "Holy Water")
+  return events.BinaryChoice(char, "Search for Holy Water?", "Yes", "No", holywater, events.Nothing())
+def Church5(char):
+  check = events.Check(char, "luck", 0)
+  loseSanity = events.Loss(char, {"sanity": 3})
```
[the remark this answers is not in the document]
Comment (ThomasLipp) [reply eldritch/encounters.py:359]: but what if I choose to stop reading the card after lose 3 sanity?

File: eldritch/test_encounters.py [hunk 13]
```diff
@@ -520,6 +624,343 @@ def testSociety4Fail(self):
...
+    choice.resolve(self.state, "No")
+    self.resolve_until_done()
+    self.assertEqual(self.char.place.name, "Society")
+
+  def testSociety7Yes(self):
+    self.state.event_stack.append(encounters.Society7(self.char))
+    choice = self.resolve_to_choice(MultipleChoice)
+    choice.resolve(self.state, "Yes")
+    choice2 = self.resolve_to_choice(CardChoice)
+    self.assertEqual(choice2.choices[0][0:9], "Rivertown")
```
[the remark this answers is not in the document]
Comment (ThomasLipp) [reply eldritch/test_encounters.py:651]: yeah, definitely better implementations, will remember in the future.

File: eldritch/encounters.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ThomasLipp) [reply eldritch/encounters.py:359]: fixed

File: eldritch/test_encounters.py [hunk 14]
```diff
@@ -520,6 +624,343 @@ def testSociety4Fail(self):
     self.assertIsNone(self.char.delayed_until)
     self.assertEqual(len(self.state.skills), 2)
 
+  def testSociety5(self):
+    self.state.event_stack.append(encounters.Society5(self.char))
+    self.resolve_until_done()
+    self.assertEqual(self.char.sanity, 2)
+
+  def testSociety5Insane(self):
```
[the remark this answers is not in the document]
Comment (ThomasLipp) [reply eldritch/test_encounters.py:632]: Understood, makes me life easier, so no objection.

File: eldritch/test_encounters.py (hunk 4 above)
Comment (ThomasLipp) [reply eldritch/test_encounters.py:940]: copy paste is still my friend

File: eldritch/encounters.py (hunk 5 above)
Comment (ThomasLipp) [reply eldritch/encounters.py:273]: is that update still in a branch, because I rebased, but I'm not seeing that.

Comment (ThomasLipp) [reply eldritch/encounters.py:273]: hmm, I see it in the base code, so I wonder why my rebase didn't work...

Comment (ThomasLipp) [reply eldritch/encounters.py:394]: Okay I think I've updated this now.  Wow..I'm moving so quickly...

File: eldritch/encounters.py (hunk 6 above)
Comment (ThomasLipp) [reply eldritch/encounters.py:414]: implemented this except for 1 dollar instead of gate trophy

File: eldritch/encounters.py (hunk 7 above)
Comment (ThomasLipp) [reply eldritch/encounters.py:464]: done

File: eldritch/encounters.py (hunk 8 above)
Comment (ThomasLipp) [reply eldritch/encounters.py:499]: updated

File: eldritch/encounters.py [hunk 15]
```diff
@@ -373,11 +373,141 @@ def Graveyard6(char):
...
+  return events.BinaryChoice(
+    char, "Spend $3 to spend the night?", "Yes", "No", stay, events.Nothing(), prereq
+  )
+
+def Church1(char):
+  return events.Loss(char, {"sanity": 1})
+def Church2(char):
+  return events.Bless(char)
+def Church3(char):
+  money = events.Loss(char, {"dollars": (char.dollars + 1)//2})
```
Comment (fingoldfish) [inline eldritch/encounters.py:477]: Probably a TODO here - the number of dollars should probably be a calculation.

File: eldritch/encounters.py [hunk 16]
```diff
@@ -373,11 +373,141 @@ def Graveyard6(char):
...
+  move = events.ForceMovement(char, "Southside")
+  loseAndMove = events.Sequence([loseSanity, move], char)
+  gainSanity = events.Gain(char, {"sanity": char.max_sanity})
+  cond =  events.Conditional(char, check, "successes", {0: loseAndMove, 1: move, 2: gainSanity})
+  return events.Sequence([check, cond], char)
+def Church6(char):
+  #TODO: remove a counter from the doomtrack
+  prereq = values.AttributePrerequisite(char, "clues", 1, "at least")
+  clue = events.Loss(char, {"clues": 1})
+  roll = events.Nothing() #TODO: roll for a success
```
Comment (fingoldfish) [inline eldritch/encounters.py:496]: You can probably just make this an events.DiceRoll

**Review by fingoldfish — APPROVED** [review]

**Pull request merged**